### PR TITLE
Remove the Identity associated type from Ledger

### PR DESCRIPTION
**Pull request by tcharding**
In an effort to remove the `Ledger` trait first remove the associated type `Identity`.  While doing this use the new `identity` modules.

Replace useages of `crate::ethereum::Address` with `identity::Ethereum`.  We have to be a bit more careful with `PublicKey` because we have a wrapper by the same name.

Works towards: #2081

### CODE REVIEW

File: cnd/src/http_api/routes/rfc003/handlers/post_swap.rs [hunk 1]
```diff
@@ -20,35 +20,35 @@ use crate::{
...
-    Facade: LoadAcceptedSwap<AL, BL, AA, BA>
-        + HtlcFunded<AL, AA>
-        + HtlcFunded<BL, BA>
-        + HtlcDeployed<AL, AA>
-        + HtlcDeployed<BL, BA>
-        + HtlcRedeemed<AL, AA>
-        + HtlcRedeemed<BL, BA>
-        + HtlcRefunded<AL, AA>
-        + HtlcRefunded<BL, BA>,
+    AI: Send + Sync + 'static + DeserializeOwned + Clone,
```
Comment (thomaseizinger) [inline cnd/src/http_api/routes/rfc003/handlers/post_swap.rs]: I would be better to apply the "say what you actually mean"-rule here. I think we just need `Accept<AI, BI>: DeserializeOwned`.

Comment (thomaseizinger) [reply cnd/src/http_api/routes/rfc003/handlers/post_swap.rs]: Also, please try to remove the `Clone` bound and specify the type you are actually calling `Clone` on.

File: cnd/src/network/mod.rs [hunk 2]
```diff
@@ -622,20 +623,22 @@ async fn handle_request(
...
     AL: Ledger,
     BL: Ledger,
     AA: Send + 'static,
     BA: Send + 'static,
-    DB: Save<Request<AL, BL, AA, BA>> + Save<Swap>,
-    Request<AL, BL, AA, BA>: Clone,
+    AI: Send + 'static,
+    BI: Send + 'static,
+    DB: Save<Request<AL, BL, AA, BA, AI, BI>> + Save<Swap>,
+    Request<AL, BL, AA, BA, AI, BI>: Send + 'static + Clone,
```
Comment (thomaseizinger) [inline cnd/src/network/mod.rs]: Shouldn't need those as we are specifying them for `AI` and `BI` above.

```suggestion
    Request<AL, BL, AA, BA, AI, BI>: Clone,
```

File: cnd/src/quickcheck.rs [hunk 3]
```diff
@@ -238,17 +237,19 @@ impl Arbitrary for Quickcheck<ledger::Ethereum> {
     }
 }
 
-impl<AL, BL, AA, BA> Arbitrary for Quickcheck<Request<AL, BL, AA, BA>>
+impl<AL, BL, AA, BA, AI, BI> Arbitrary for Quickcheck<Request<AL, BL, AA, BA, AI, BI>>
 where
     AL: Ledger,
```
Comment (thomaseizinger) [inline cnd/src/quickcheck.rs]: I think these bounds can go away now. Please also check in other places, where we previously only set them to access `L::Identity`.

File: cnd/src/swap_protocols/rfc003/actions/bitcoin.rs [hunk 4]
```diff
@@ -13,14 +13,14 @@ use crate::{
 use ::bitcoin::{Amount, OutPoint, Transaction};
 use blockchain_contracts::bitcoin::{rfc003::bitcoin_htlc::BitcoinHtlc, witness::PrimedInput};
 
-impl<B> FundAction<B, asset::Bitcoin> for (B, asset::Bitcoin)
+impl<B> FundAction<B, asset::Bitcoin, identity::Bitcoin> for (B, asset::Bitcoin, identity::Bitcoin)
```
Comment (thomaseizinger) [inline cnd/src/swap_protocols/rfc003/actions/bitcoin.rs]: Why do we need to expand the tuple here?

```suggestion
impl<B> FundAction<B, asset::Bitcoin, identity::Bitcoin> for (B, asset::Bitcoin)
```

The reason this is implemented for a tuple is so we can use `where` clauses with this tuple in other places (like here: https://github.com/comit-network/comit-rs/blob/ede13873527e377e2c9707e6a4a472f968624963/cnd/src/swap_protocols/rfc003/bob/actions/erc20.rs#L19). That is only useful for completely orthogonal parameters (ledgers and assets). The identity is always bound to the ledger and hence shouldn't need to be included in this tuple.

File: cnd/src/swap_protocols/rfc003/actions/ether.rs [hunk 5]
```diff
@@ -14,11 +14,13 @@ use crate::{
 };
 use blockchain_contracts::ethereum::rfc003::ether_htlc::EtherHtlc;
 
-impl FundAction<Ethereum, asset::Ether> for (Ethereum, asset::Ether) {
+impl FundAction<Ethereum, asset::Ether, identity::Ethereum>
+    for (Ethereum, asset::Ether, identity::Ethereum)
```
Comment (thomaseizinger) [inline cnd/src/swap_protocols/rfc003/actions/ether.rs]: Same here, we shouldn't need this:
```suggestion
    for (Ethereum, asset::Ether)
```

File: cnd/src/swap_protocols/rfc003/actions/mod.rs [hunk 6]
```diff
@@ -21,37 +21,37 @@ pub enum Action<Accept, Decline, Deploy, Fund, Redeem, Refund> {
     Refund(Refund),
 }
 
-pub trait FundAction<L, A>
+pub trait FundAction<L, A, I>
 where
     L: Ledger,
```
Comment (thomaseizinger) [inline cnd/src/swap_protocols/rfc003/actions/mod.rs]: I think you should be able to remove this bound now.

File: cnd/src/swap_protocols/rfc003/ethereum/mod.rs [hunk 7]
```diff
@@ -42,7 +42,6 @@ impl From<Seconds> for Duration {
 
 impl Ledger for Ethereum {
     type HtlcLocation = Address;
-    type Identity = Address;
```
Comment (thomaseizinger) [inline cnd/src/swap_protocols/rfc003/ethereum/mod.rs]: :tada:

**Comment by thomaseizinger** [comment]
@tcharding I am fine with these changes. If you are too, feel free to merge this so we can move forward!

**Comment by tcharding** [comment]
Ha!  I missed the best bit, removing the actual `Ledger` trait bounds off all the impls.  Thanks @thomaseizinger.  Also I clearly didn't fully internalise your finding of 'say just what you mean' :)

**Pull request merged**